File: src/cpp/OrthogonalPolynomials.cpp
/*
  Class to evaluate orthogonal polynomials.

  Author: Komahan Boopathy (komahanboopathy@gmail.com)
*/

#include <stdio.h>
#include <math.h>

#include "OrthogonalPolynomials.h"

/**
   Constructor for orthogonal polynomials
*/
OrthogonalPolynomials::OrthogonalPolynomials(){}

/**
   Destructor for orthogonal polynomials
*/
OrthogonalPolynomials::~OrthogonalPolynomials(){}

/**
   Evaluate the Hermite polynomials

   @param z the point to evaluate the basis
   @param d the degree of basis function
*/
scalar OrthogonalPolynomials::hermite(scalar z, int d){
  if ( d <= 4 ) {
    return explicit_hermite(z,d);
  } else {
    return recursive_hermite(z,d);
  }
}

/**
   Evaluate the unit Hermite polynomials

   @param z the point to evaluate the basis
   @param d the degree of basis function
*/
scalar OrthogonalPolynomials::unit_hermite(scalar z, int d){
  return hermite(z,d)/sqrt(factorial(d));
}

/**
   Evaluate the Laguerre polynomials

   @param z the point to evaluate the basis
   @param d the degree of basis function
*/
scalar OrthogonalPolynomials::laguerre(scalar z, int d){
  if ( d <= 4 ) {
    return explicit_laguerre(z,d);
  } else {
    return recursive_laguerre(z,d);
  }
}

/**
   Evaluate unit Laguerre polynomials (already normalized)

   @param z the point to evaluate the basis
   @param d the degree of basis function
*/
scalar OrthogonalPolynomials::unit_laguerre(scalar z, int d){
  return laguerre(z,d);
}

/**
   Evaluate Legendre polynomials

   @param z the point to evaluate the basis
   @param d the degree of basis function
*/
scalar OrthogonalPolynomials::legendre(scalar z, int d){
  if ( d <= 4 ) {
    return explicit_legendre(z,d);
  } else {
    return general_legendre(z,d);
  }
}

/**
   Evaluate unit legendre polynomials

   @param z the point to evaluate the basis
   @param d the degree of basis function
*/
scalar OrthogonalPolynomials::unit_legendre(scalar z, int d){
  return legendre(z,d)*sqrt(scalar(2*d+1));
}

/**
   Combination nCr = n!/((n-r)!r!)
*/
scalar OrthogonalPolynomials::comb(int n, int r){
  scalar nfact  = factorial(n);
  scalar rfact  = factorial(r);
  scalar nrfact = factorial(n-r);
  return nfact/(rfact*nrfact);
}

/**
   Compute the factorial of a number

   @param n the numbr for which factorial is needed
*/
scalar OrthogonalPolynomials::factorial( int n ){
  scalar factorial;
  if ( n == 0 ){
    factorial = 1.0;
  } else if ( n == 1 ){
    factorial = 1.0;
  } else if ( n == 2 ){
    factorial = 2.0;
  } else if ( n == 3 ){
    factorial = 6.0;
  } else if ( n == 4 ){
    factorial = 24.0;
  } else if ( n == 5 ){
    factorial = 120.0;
  } else if ( n == 6 ){
    factorial = 720.0;
  } else if ( n == 7 ){
    factorial = 5040.0;
  } else if ( n == 8 ){
    factorial = 40320.0;
  } else if ( n == 9 ){
    factorial = 362880.0;
  } else if ( n == 10 ){
    factorial = 3628800.0;
  } else {
    factorial = 1.0;
    for ( int i = 1; i <= n; i++ ) {
      factorial *= scalar(i);
    }
  }
  return factorial;
}

/**
   Hermite polynomials are evaluated using explicit expressions

   @param z the point to evaluate the basis
   @param d the degree of basis function
*/
scalar OrthogonalPolynomials::explicit_hermite(scalar z, int d){
  scalar hval = 0.0;
  if ( d == 0 ){
    hval = 1.0;
  } else if ( d == 1 ){
    hval = z;
  } else if ( d == 2 ){
    hval = z*z - 1.0;
  } else if ( d == 3 ){
    hval = z*z*z -3.0*z;
  } else if ( d == 4 ){
    hval = z*z*z*z -6.0*z*z + 3.0;
  }
  return hval;
}

/**
   Hermite polynomials are evaluated using recursive expressions

   @param z the point to evaluate the basis
   @param d the degree of basis function
*/
scalar OrthogonalPolynomials::recursive_hermite(scalar z, int d){
  scalar hval = 0.0;
  if ( d == 0 ) {
    hval = 1.0;
  } else if ( d == 1 ) {
    hval = z;
  } else {
    hval = z*recursive_hermite(z,d-1) - scalar(d-1)*recursive_hermite(z,d-2);
  }
  return hval;
}

/**
   Laguerre polynomials are evaluated using explicit expressions

   @param z the point to evaluate the basis
   @param d the degree of basis function
*/
scalar OrthogonalPolynomials::explicit_laguerre(scalar z, int d){
  scalar lval = 0.0;
  if ( d == 0 ){
    lval = 1.0;
  } else if ( d == 1 ){
    lval = 1.0 - z;
  } else if ( d == 2 ){
    lval = z*z - 4.0*z + 2.0;
    lval /= factorial(2);
  } else if ( d == 3 ){
    lval = -z*z*z + 9.0*z*z - 18.0*z + 6.0;
    lval /= factorial(3);
  } else if ( d == 4 ){
    lval = z*z*z*z - 16.0*z*z*z + 72.0*z*z - 96.0*z + 24.0;
    lval /= factorial(4);
  }
  return lval;
}

/**
   Laguerre polynomials are evaluated using recursive expressions

   @param z the point to evaluate the basis
   @param d the degree of basis function
*/
scalar OrthogonalPolynomials::recursive_laguerre(scalar z, int d){
  scalar lval = 0.0;
  if ( d == 0 ) {
    lval = 1.0;
  } else if ( d == 1 ) {
    lval = 1.0 - z;
  } else {
    lval = ((scalar(2*d-1)-z)*recursive_laguerre(z,d-1) - scalar(d-1)*recursive_laguerre(z,d-2));
    lval /= scalar(d);
  }
  return lval;
}

/**
   Legendre polynomials are evaluated using explicit expressions

   @param z the point to evaluate the basis
   @param d the degree of basis function
*/
scalar OrthogonalPolynomials::explicit_legendre(scalar z, int d){
  scalar pval = 0.0;
  if ( d == 0 ){
    pval = 1.0;
  } else if ( d == 1 ){
    pval = 2.0*z - 1.0;
  } else if ( d == 2 ){
    pval = 6.0*z*z - 6.0*z + 1.0;
  } else if ( d == 3 ){
    pval = 20.0*z*z*z - 30.0*z*z + 12.0*z - 1.0;
  } else if ( d == 4 ){
    pval = 70.0*z*z*z*z - 140.0*z*z*z + 90.0*z*z - 20.0*z + 1.0;
  }
  return pval;
}

/**
   Legendre polynomials are evaluated using general expressions

   @param z the point to evaluate the basis
   @param d the degree of basis function
*/
scalar OrthogonalPolynomials::general_legendre(scalar z, int d){
  scalar pval = 0.0;
  for (int k = 0; k <= d; k++){
    pval = pval + comb(d,k)*comb(d+k,k)*pow(-z,k);
  }
  pval = pval*pow(-1.0,d);
  return pval;
}

/**
   Function to evaluate polynomials
*/
void test_polynomials( int argc, char *argv[] ){

  OrthogonalPolynomials *poly = new OrthogonalPolynomials();

  scalar z = 1.1;
  int max_order = 10;
  int nruns = 100000;
  for (int j = 0; j < nruns; j++){
    // Test hermite polynomials
    printf("hermite\n");
    for (int i = 0; i < max_order; i++){
      printf("%2d %10.3f %10.3f \n", i, RealPart(poly->hermite(z,i)), RealPart(poly->unit_hermite(z,i)));
    }

    // Test Legendre polynomials
    printf("\nLegendre\n");
    for (int i = 0; i < max_order; i++){
      printf("%2d %10.3f %10.3f \n", i, RealPart(poly->legendre(z,i)), RealPart(poly->unit_legendre(z,i)));
    }

    // Test Laguerre polynomials
    printf("\nLaguerre\n");
    for (int i = 0; i < max_order; i++){
      printf("%2d %10.3f %10.3f \n", i, RealPart(poly->laguerre(z,i)), RealPart(poly->unit_laguerre(z,i)));
    }
  }
  delete poly;
}

File: src/fortran/orthogonal_polynomials.f90
module orthogonal_polynomials

  implicit none

  !private
  !public :: hermite , unit_hermite
  !public :: laguerre, unit_laguerre
  !public :: legendre, unit_legendre

contains

  !===================================================================!
  ! Compute the factorial of given integer number
  !===================================================================!

  pure function factorial(n)

    integer, intent(in) :: n
    real(8) :: factorial
    integer :: i
    
    if ( n .eq. 0 ) then
       factorial = 1.0d0
    else if ( n .eq. 1 ) then
       factorial = 1.0d0
    else if ( n .eq. 2 ) then   
       factorial = 2.0d0       
    else if ( n .eq. 3 ) then
       factorial = 6.0d0              
    else if ( n .eq. 4 ) then
       factorial = 24.0d0                     
    else if ( n .eq. 5 ) then
       factorial = 120.0d0                            
    else if ( n .eq. 6 ) then
       factorial = 720.0d0                                          
    else if ( n .eq. 7 ) then
       factorial = 5040.0d0                                          
    else if ( n .eq. 8 ) then
       factorial = 40320.0d0                                                        
    else if ( n .eq. 9 ) then
       factorial = 362880.0d0
    else if ( n .eq. 10 ) then
       factorial = 3628800.0d0
    else    
       factorial = 1.0d0
       do i = 1, n
          factorial = factorial*dble(i)
       end do
    end if

  end function factorial

  !===================================================================!
  ! Combination nCr = n!/((n-r)!r!)
  !===================================================================!
  
  pure function comb(n,r)

    integer, intent(in) :: n, r
    real(8) :: comb
    real(8) :: nfact, rfact, nrfact
    
    nfact  = factorial(n)
    rfact  = factorial(r)
    nrfact = factorial(n-r)

    comb = nfact/(rfact*nrfact)
    
  end function comb
  
  !===================================================================!
  ! Hermite polynomial of degree d, evaluated at z: H(z,d)
  !===================================================================!

  pure recursive function recursive_hermite(z, d) result(hval)

    real(8), intent(in) :: z
    integer, intent(in) :: d
    real(8) :: hval

    ! Use two term recursion formulae
    if (d == 0) then
       hval =  1.0d0
    else if (d == 1) then
       hval = z
    else 
       hval = z*recursive_hermite(z,d-1) - dble(d-1)*recursive_hermite(z,d-2)
    end if

  end function recursive_hermite

  pure function explicit_hermite(z, d) result(hval)

    real(8), intent(in) :: z
    integer, intent(in) :: d
    real(8) :: hval

    ! Use two term recursion formulae
    if (d .eq. 0) then
       hval =  1.0d0
    else if (d .eq. 1) then
       hval = z
    else if (d .eq. 2) then
       hval = z**2 - 1.0d0
    else if (d .eq. 3) then
       hval = z**3 - 3.0d0*z
    else if (d .eq. 4) then
       hval = z**4 - 6.0d0*z*z + 3.0d0
    end if

  end function explicit_hermite

  pure function hermite(z, d)

    real(8), intent(in) :: z
    integer, intent(in) :: d
    real(8) :: hermite

    if (d .le. 4) then 
       hermite = explicit_hermite(z, d)
    else
       hermite = recursive_hermite(z, d)
    end if

  end function hermite

  !===================================================================!
  ! Normalize the Hermite polynomial \hat{H}(z,d)
  !===================================================================!

  pure function unit_hermite(z,d)

    real(8), intent(in) :: z
    integer, intent(in) :: d
    real(8) :: unit_hermite

    unit_hermite = hermite(z,d)/sqrt(factorial(d))

  end function unit_hermite

  !===================================================================!
  ! Laguerre polynomial of degree d, evaluated at z: L(z,d)
  !===================================================================!
  
  pure recursive function recursive_laguerre(z, d) result(lval)

    real(8), intent(in) :: z
    integer, intent(in) :: d
    real(8) :: lval

    ! Use two term recursion formulae
    if (d == 0) then
       lval = 1.0d0
    else if (d == 1) then
       lval = 1.0d0 - z  
    else 
       lval = ((dble(2*d-1)-z)*recursive_laguerre(z,d-1) &
            & - dble(d-1)*recursive_laguerre(z,d-2))/dble(d)
    end if

  end function recursive_laguerre
 
  pure recursive function explicit_laguerre(z, d) result(lval)

    real(8), intent(in) :: z
    integer, intent(in) :: d
    real(8) :: lval

    ! Use two term recursion formulae
    if (d .eq. 0) then
       lval = 1.0d0
    else if (d .eq. 1) then
       lval = 1.0d0 - z
    else if (d .eq. 2) then
       lval = z**2 - 4.0d0*z + 2.0d0
       lval = lval/factorial(2)
    else if (d .eq. 3) then
       lval = -z**3 + 9.0d0*z**2 - 18.0d0*z + 6.0d0
       lval = lval/factorial(3)
    else if (d .eq. 4) then
       lval = z**4 - 16.0d0*z**3 + 72.0d0*z**2 - 96.0d0*z + 24.0d0
       lval = lval/factorial(4)
    end if
    
  end function explicit_laguerre

  pure function laguerre(z, d)

    real(8), intent(in) :: z
    integer, intent(in) :: d
    real(8) :: laguerre

    if (d .le. 4) then 
       laguerre = explicit_laguerre(z, d)
    else
       laguerre = recursive_laguerre(z, d)
    end if

  end function laguerre

  !===================================================================!
  ! Normalize the Laguerre polynomial \hat{L}(z,d)
  !===================================================================!

  pure function unit_laguerre(z,d)

    real(8), intent(in) :: z
    integer, intent(in) :: d
    real(8) :: unit_laguerre

    unit_laguerre = laguerre(z,d)

  end function unit_laguerre

  
  !===================================================================!
  ! Legendre polynomial of degree d, evaluated at z: P(z,d)
  !===================================================================!
  
  pure recursive function general_legendre(z, d) result(pval)

    real(8), intent(in) :: z
    integer, intent(in) :: d
    integer :: k
    real(8) :: pval

    ! using general formula
    pval = 0.0d0
    do k = 0, d
       pval = pval + comb(d,k)*comb(d+k,k)*(-z)**k
    end do
    pval = pval*(-1.0d0)**d       

  end function general_legendre

  pure recursive function explicit_legendre(z, d) result(pval)

    real(8), intent(in) :: z
    integer, intent(in) :: d
    real(8) :: pval

    ! Use two term recursion formulae
    if (d .eq. 0) then
       pval = 1.0d0
    else if (d .eq. 1) then
       pval = 2.0d0*z - 1.0d0
    else if (d .eq. 2) then
       pval = 6.0d0*z**2 - 6.0d0*z + 1
    else if (d .eq. 3) then
       pval = 20.0d0*z**3 -30.0d0*z**2 + 12.0d0*z - 1.0d0
    else if (d .eq. 4) then
       pval = 70.0d0*z**4 - 140.0d0*z**3 + 90.0d0*z**2 - 20.0d0*z + 1.0d0
    end if

  end function explicit_legendre

  pure function legendre(z, d)

    real(8), intent(in) :: z
    integer, intent(in) :: d
    real(8) :: legendre

    if (d .le. 4) then 
       legendre = explicit_legendre(z, d)
    else
       legendre = general_legendre(z, d)
    end if

  end function legendre

  !===================================================================!
  ! Normalize the Legendre polynomial \hat{L}(z,d)
  !===================================================================!

  pure function unit_legendre(z,d)

    real(8), intent(in) :: z
    integer, intent(in) :: d
    real(8) :: unit_legendre

    unit_legendre = legendre(z,d)*sqrt(dble(2*d+1))

  end function unit_legendre
  
end module orthogonal_polynomials

!!$program test_orthonormal_polynomials
!!$
!!$  use orthogonal_polynomials
!!$  implicit none
!!$
!!$  integer :: i, n, j
!!$  real(8), parameter :: z = 1.1d0
!!$  integer, parameter :: max_order = 9, nruns = 100000
!!$  real(8) :: a
!!$
!!$  do n = 1, 10000
!!$     do i = 0, max_order
!!$        a = hermite(z,i)
!!$ !       print *, i, legendre(z,i)
!!$ !       print *, i, laguerre(z,i)             
!!$     end do
!!$  end do
!!$
!!$  print *, a
!!$
!!$  do j = 1, nruns
!!$  print *, "hermite"
!!$  do i = 0, max_order
!!$     write(*,"(i2,F10.3,F10.3)") i, hermite(z,i), unit_hermite(z,i)
!!$  end do
!!$
!!$  print *, ""
!!$  print *, "legendre"
!!$  do i = 0, max_order
!!$     write(*,"(i2,F10.3,F10.3)") i, legendre(z,i), unit_legendre(z,i)
!!$  end do
!!$
!!$  print *, ""
!!$  print *, "laguerre"
!!$  do i = 0, max_order
!!$     write(*,"(i2,F10.3,F10.3)") i, laguerre(z,i), unit_laguerre(z,i)
!!$  end do
!!$  end do
!!$
!!$end program test_orthonormal_polynomials
